gtsfm/evaluation/metrics.py

```diff
@@ -0,0 +1,362 @@
+                     Has to be provided if data = None.
+            store_full_data: Whether all the values have to be stored and saved or only summary is required. True by default.
+            plot_type: The plot to use for visualization of the metric. 
+                       Defaults:
+                          PlotType.BAR if data is a scalar
+                          PlotType.BOX if data is a distribution (other option is PlotType.HISTOGRAM)
+                        It is inferred from the summary if plot_type is not provided and summary is.
+    """
+        if summary is None and data is None:
+            raise ValueError("Data and summary cannot both be None.")
+
+        self._name = name
+        if data is not None:
+            # Cast to a numpy array
+            if not isinstance(data, np.ndarray):
+                data = np.array(data)
+            if data.ndim > 1:
+                raise ValueError("Metrics must be scalars on 1D-distributions.")
+
+            # Save dimension and plot_type for data
+            self._dim = data.ndim
+            plot_types_for_dim = self._get_plot_types_for_dim(self._dim)
+            if plot_type is None:
+                if summary is not None:
+                    self._plot_type = self.PlotType.HISTOGRAM if "histogram" in summary else self.PlotType.BOX
+                else:
+                    self._plot_type = plot_types_for_dim[0]
+            elif plot_type in plot_types_for_dim:
+                self._plot_type = plot_type
+            else:
+                raise ValueError("Unsupported plot type for the data dimension")
+
+            # Create a summary if the data is a 1D distribution
+            if self._dim == 1:
+                self._summary = self._create_summary(data)
+
+            # Store full data only if its a scalar or if asked to.
+            if self._dim == 0 or store_full_data:
+                self._data = data
+            else:
+                self._data = None
+        else:
+            self._dim = 1
+            self._summary = summary
+            self._plot_type = self.PlotType.HISTOGRAM if "histogram" in summary else self.PlotType.BOX
+            self._data = None
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def data(self) -> np.array:
+        return self._data
+
+    @property
+    def plot_type(self) -> PlotType:
+        return self._plot_type
+
+    @property
+    def dim(self) -> int:
+        return self._dim
+
+    @property
+    def summary(self) -> Dict[str, Any]:
+        return self._summary
+
+    def _get_plot_types_for_dim(self, dim: int) -> List[PlotType]:
+        if dim == 0:
+            return [self.PlotType.BAR]
+        if dim == 1:
+            return [self.PlotType.BOX, self.PlotType.HISTOGRAM]
+        return []
+
+    def _get_distribution_histogram(self, data: np.ndarray) -> Dict[str, Union[float, int]]:
```

Comment:
this does not need to be a function in the class.

Reply:
agreed, i would prefer this to be in this file as a free function

Reply:
But then users could import it, which we dont want.

Reply:
The same argument could be used for all internal methods in this class, I made them so because I did not want them to be imported without the class.

Reply:
this fn doesnt access any private class attributes, though, right? i think we prefer functional when possible, since it makes unit testing so much easier

Reply:
Even if we are not going to write a unit test for that function? We are not quite violating the functional paradigm if we have internal methods that do not change any data member of the class. This is still functional, the only difference is where we have the function. Since its not going to be used externally I thought it makes sense to have it as an internal member with an _ in the name indicating this. 
It also makes semantic sense, as these functions are not meaningful outside the context of this class.

Reply:
While its not my preference, I dont mind moving these functions outside the class. I would move _get_distribution_histogram, _get_distribution_quartiles, and _create_summary, since they are all similar. Let me know if you disagree with my motivation above. :)

Reply:
i prefer just this one function as a free function, since it doesnt use any aspect of the class.

unit test would be nice too : - )

Reply:
Moved to free function, current usage is covered by the tests in the class, so have not added more tests. The only untested behavior is the histogram for float data, which is not used by any metric currently.